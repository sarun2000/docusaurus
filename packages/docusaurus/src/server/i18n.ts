/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import {I18n, DocusaurusConfig, I18nLocaleConfig} from '@docusaurus/types';
import path from 'path';
import {normalizeUrl} from '@docusaurus/utils';

export function defaultLocaleConfig(locale: string): I18nLocaleConfig {
  return {
    label: locale,
  };
}

export async function loadI18n(
  config: DocusaurusConfig,
  options: {locale?: string} = {},
): Promise<I18n> {
  const i18nConfig = config.i18n;
  const currentLocale = options.locale ?? i18nConfig.defaultLocale;

  if (currentLocale && !i18nConfig.locales.includes(currentLocale)) {
    throw new Error(
      `It is not possible to load Docusaurus with locale="${currentLocale}".
This locale is not in the available locales of your site configuration: config.i18n.locales=[${i18nConfig.locales.join(
        ',',
      )}]
Note: Docusaurus only support running one local at a time.`,
    );
  }

  function getLocaleConfig(locale: string): I18nLocaleConfig {
    // User provided values
    const localeConfigOptions: Partial<I18nLocaleConfig> =
      i18nConfig.localeConfigs[locale];

    return {...defaultLocaleConfig(locale), ...localeConfigOptions};
  }

  const localeConfigs = i18nConfig.locales.reduce((acc, locale) => {
    return {...acc, [locale]: getLocaleConfig(locale)};
  }, {});

  return {
    ...i18nConfig,
    currentLocale,
    localeConfigs,
  };
}

export function localizePath({
  pathType,
  path: originalPath,
  i18n,
  options = {},
}: {
  pathType: 'fs' | 'url';
  path: string;
  i18n: I18n;
  options?: {localizePath?: boolean};
}): string {
  const shouldLocalizePath: boolean =
    typeof options.localizePath === 'undefined'
      ? // By default, we don't localize the path of defaultLocale
        i18n.currentLocale !== i18n.defaultLocale
      : options.localizePath;

  if (shouldLocalizePath) {
    // FS paths need special care, for Windows support
    if (pathType === 'fs') {
      return path.join(originalPath, path.sep, i18n.currentLocale, path.sep);
    }
    // Url paths
    else if (pathType === 'url') {
      return normalizeUrl([originalPath, '/', i18n.currentLocale, '/']);
    }
    // should never happen
    else {
      throw new Error(`unhandled pathType=${pathType}`);
    }
  } else {
    return originalPath;
  }
}
