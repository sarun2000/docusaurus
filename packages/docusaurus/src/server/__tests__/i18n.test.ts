/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {loadI18n, localizePath, defaultLocaleConfig} from '../i18n';
import {DEFAULT_I18N_CONFIG} from '../configValidation';
import path from 'path';
import {chain, identity} from 'lodash';

function testLocaleConfigsFor(locales: string[]) {
  return chain(locales).keyBy(identity).mapValues(defaultLocaleConfig).value();
}

describe('loadI18n', () => {
  test('should load I18n for default config', async () => {
    await expect(
      loadI18n(
        // @ts-expect-error: enough for this test
        {
          i18n: DEFAULT_I18N_CONFIG,
        },
      ),
    ).resolves.toEqual({
      defaultLocale: 'en',
      locales: ['en'],
      currentLocale: 'en',
      localeConfigs: testLocaleConfigsFor(['en']),
    });
  });

  test('should load I18n for multi-lang config', async () => {
    await expect(
      loadI18n(
        // @ts-expect-error: enough for this test
        {
          i18n: {
            defaultLocale: 'fr',
            locales: ['en', 'fr', 'de'],
            localeConfigs: {},
          },
        },
      ),
    ).resolves.toEqual({
      defaultLocale: 'fr',
      locales: ['en', 'fr', 'de'],
      currentLocale: 'fr',
      localeConfigs: testLocaleConfigsFor(['en', 'fr', 'de']),
    });
  });

  test('should load I18n for multi-locale config with specified locale', async () => {
    await expect(
      loadI18n(
        // @ts-expect-error: enough for this test
        {
          i18n: {
            defaultLocale: 'fr',
            locales: ['en', 'fr', 'de'],
            localeConfigs: {},
          },
        },
        {locale: 'de'},
      ),
    ).resolves.toEqual({
      defaultLocale: 'fr',
      locales: ['en', 'fr', 'de'],
      currentLocale: 'de',
      localeConfigs: testLocaleConfigsFor(['en', 'fr', 'de']),
    });
  });

  test('should load I18n for multi-locale config with some xcustom locale configs', async () => {
    await expect(
      loadI18n(
        {
          i18n: {
            defaultLocale: 'fr',
            locales: ['en', 'fr', 'de'],
            localeConfigs: {
              fr: {label: 'Français'},
              // @ts-expect-error: empty on purpose
              en: {},
            },
          },
        },
        {locale: 'de'},
      ),
    ).resolves.toEqual({
      defaultLocale: 'fr',
      locales: ['en', 'fr', 'de'],
      currentLocale: 'de',
      localeConfigs: {
        fr: {label: 'Français'},
        en: defaultLocaleConfig('en'),
        de: defaultLocaleConfig('de'),
      },
    });
  });

  test('should throw when trying to load undeclared locale', async () => {
    await expect(
      loadI18n(
        // @ts-expect-error: enough for this test
        {
          i18n: {
            defaultLocale: 'fr',
            locales: ['en', 'fr', 'de'],
            localeConfigs: {},
          },
        },
        {locale: 'it'},
      ),
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});

describe('localizePath', () => {
  test('should localize url path with current locale', () => {
    expect(
      localizePath({
        pathType: 'url',
        path: '/baseUrl',
        i18n: {
          defaultLocale: 'en',
          locales: ['en', 'fr'],
          currentLocale: 'fr',
          localeConfigs: {},
        },
        options: {localizePath: true},
      }),
    ).toEqual('/baseUrl/fr/');
  });

  test('should localize fs path with current locale', () => {
    expect(
      localizePath({
        pathType: 'fs',
        path: '/baseFsPath',
        i18n: {
          defaultLocale: 'en',
          locales: ['en', 'fr'],
          currentLocale: 'fr',
          localeConfigs: {},
        },
        options: {localizePath: true},
      }),
    ).toEqual(`/baseFsPath${path.sep}fr${path.sep}`);
  });

  test('should localize path for default locale, if requested', () => {
    expect(
      localizePath({
        pathType: 'url',
        path: '/baseUrl/',
        i18n: {
          defaultLocale: 'en',
          locales: ['en', 'fr'],
          currentLocale: 'en',
          localeConfigs: {},
        },
        options: {localizePath: true},
      }),
    ).toEqual('/baseUrl/en/');
  });

  test('should not localize path for default locale by default', () => {
    expect(
      localizePath({
        pathType: 'url',
        path: '/baseUrl/',
        i18n: {
          defaultLocale: 'en',
          locales: ['en', 'fr'],
          currentLocale: 'en',
          localeConfigs: {},
        },
        // options: {localizePath: true},
      }),
    ).toEqual('/baseUrl/');
  });

  test('should localize path for non-default locale by default', () => {
    expect(
      localizePath({
        pathType: 'url',
        path: '/baseUrl/',
        i18n: {
          defaultLocale: 'en',
          locales: ['en', 'fr'],
          currentLocale: 'en',
          localeConfigs: {},
        },
        // options: {localizePath: true},
      }),
    ).toEqual('/baseUrl/');
  });
});
