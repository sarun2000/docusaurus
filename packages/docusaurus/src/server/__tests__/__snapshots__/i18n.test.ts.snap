// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadI18n should throw when trying to load undeclared locale 1`] = `
"It is not possible to load Docusaurus with locale=\\"it\\".
This locale is not in the available locales of your site configuration: config.i18n.locales=[en,fr,de]
Note: Docusaurus only support running one local at a time."
`;
